## Advanced Deep Learning with Keras in Python

<img src="images/1.PNG" style="width:950px;height:400px;"> 

## 1 The Keras Functional API

<img src="images/2.PNG" style="width:950px;height:400px;"> 

## 2 Two Input Networks Using Categorical Embeddings, Shared Layers, and Merge Layers 

<img src="images/3.PNG" style="width:950px;height:500px;"> 

## 3 Multiple Inputs: 3 Inputs (and Beyond!)

<img src="images/4.PNG" style="width:950px;height:500px;"> 

## 4 Multiple Outputs

<img src="images/5.PNG" style="width:950px;height:500px;"> 

<img src="images/6.PNG" style="width:950px;height:500px;"> 

## Keras input and dense layers

<img src="images/7.PNG" style="width:650px;height:300px;"> 

<img src="images/8.PNG" style="width:650px;height:300px;"> 

<img src="images/9.PNG" style="width:650px;height:300px;"> 

## Inputs

<img src="images/10.PNG" style="width:650px;height:300px;"> 

   - The number of columns in the input is specified using the shape parameter.
   - This tells the model how much data to expect.
   - The shape arugment expects a tuple.

<img src="images/11.PNG" style="width:650px;height:300px;"> 

## Output

<img src="images/12.PNG" style="width:650px;height:300px;"> 

   - Outputs in keras are most commonly a single dense layer, which specifies the shape of the expected output.
   - In this case, we are expecting our model to predict a single value, so we pass one unit to the dense layer.

## The difference between layers and tensors is key to understand the Keras function API.
   - Layers are used to construct a deep learning model.
   - Tensor are used to define the data flow through the model.

<img src="images/13.PNG" style="width:650px;height:300px;"> 

   - Input layer defines a tensor, which we pass to the output_layer() function.  
   - The final output of our model is tensor.

<img src="images/14.PNG" style="width:650px;height:300px;"> 

## Input layers

 - The first step in creating a neural network model is to define the Input layer. This layer takes in raw data, usually in the form of numpy arrays. The shape of the Input layer defines how many variables your neural network will use. For example, if the input data has 10 columns, you define an Input layer with a shape of (10,).
 
In this case, you are only using one input in your network.
 
## Instructions
  - Import the Input layer function from keras.layers.
  - Create an input layer of shape 1.


In [4]:
## Code 
# Import Input from keras.layers
from keras.layers import Input

# Create an input layer of shape 1
input_tensor = Input(shape=(1,))

Using TensorFlow backend.


## Remember that the input layer allows your model to load data.

## Dense layers

   - Once you have an Input layer, the next step is to add a Dense layer.

   - Dense layers learn a weight matrix, where the first dimension of the matrix is the dimension of the input data, and the second dimension is the dimension of the output data. Recall that your Input layer has a shape of 1. In this case, your output layer will also have a shape of 1. This means that the Dense layer will learn a 1x1 weight matrix.

In this exercise, you will add a dense layer to your model, after the input layer.

## Instructions

 - Import the Dense layer function from keras.layers.
 - Create a Dense layer with 1 unit.
 - Pass input_tensor to output_layer()

In [1]:
# Load layers
from keras.layers import Input, Dense

# Input layer
input_tensor = Input(shape=(1,))

# Dense layer
output_layer = Dense(1)

# Connect the dense layer to the input_tensor
output_tensor = output_layer(input_tensor)

Using TensorFlow backend.


## This network will take the input, apply a linear coefficient to it, and return the result

## Output layers

 - Output layers are simply Dense layers! 
 - Output layers are used to reduce the dimension of the inputs to the dimension of the outputs. 
 - You'll learn more about output dimensions in chapter 4, but for now, you'll always use a single output in your neural networks, which is equivalent to Dense(1) or a dense layer with a single unit.
 
 
## Instructions


  - Import the Input and Dense functions from keras.layers.
  - Create an input layer of shape 1.
  - Again, create a dense layer with 1 unit and pass input_tensor directly to it.


In [2]:
# Load layers
from keras.layers import Input, Dense

# Input layer
input_tensor = Input(shape=(1,))

# Create a dense layer and connect the dense layer to the input_tensor in one step
# Note that we did this in 2 steps in the previous exercise, but are doing it in one step now
output_tensor = Dense(1)(input_tensor)

 - One line to both create the layer function and then call it to produce a tensor.

## The output layer allows your model to make predictions.

## Build and compile a model

<img src="images/15.PNG" style="width:650px;height:300px;"> 

 - In this case, we only have a single input and a single output, which we directly pass to the mode
 - In case of mulitiple inputs an multiple outputs you will pass lists of inputs or list of outputs to the model.

## Compile a model

<img src="images/16.PNG" style="width:650px;height:300px;"> 

## Summarize the model

<img src="images/17.PNG" style="width:650px;height:300px;"> 

## Plot model Keras

<img src="images/18.PNG" style="width:850px;height:500px;"> 

## Build a model

 - Once you've defined an input layer and an output layer, you can build a Keras model. The model object is how you tell Keras where the model starts and stops: where data comes in and where predictions come out.
 
## Instructions
 - Import Model from keras.models to create a keras model.
 - Use the input layer and output layer you already defined as the model's input and output.


In [3]:
# Input/dense/output layers
from keras.layers import Input, Dense
input_tensor = Input(shape=(1,))
output_tensor = Dense(1)(input_tensor)

# Build the model
from keras.models import Model
model = Model(input_tensor, output_tensor)

## Compile a model

 - The final step in creating a model is compiling it. Now that you've created a model, you have to compile it before you can fit it to data. This finalizes your model, freezes all its settings, and prepares it to meet some data!

 - During compilation, you specify the optimizer to use for fitting the model to the data, and a loss function. 'adam' is a good default optimizer to use, and will generally work well. Loss function depends on the problem at hand. Mean squared error is a common loss function and will optimize for predicting the mean, as is done in least squares regression.

 - Mean absolute error optimizes for the median and is used in quantile regression. For this dataset, 'mean_absolute_error' works pretty well, so use it as your loss function.
 

## Instructions
  - Compile the model you created (model).
  - Use the 'adam' optimizer.
  - Use mean absolute error (or 'mean_absolute_error') loss.


In [4]:
# Compile the model
model.compile(optimizer='adam', loss='mean_absolute_error')

## Visualize a model

 - Now that you've compiled the model, take a look a the result of your hard work! You can do this by looking at the model summary, as well as its plot.

 - The summary will tell you the names of the layers, as well as how many units they have and how many parameters are in the model.

 - The plot will show how the layers connect to each other.
 
 
## Instructions
  - Summarize the model.
  - Plot the model.


## pip install pydot

In [44]:
import pydot
from keras.utils.vis_utils import model_to_dot
keras.utils.vis_utils.pydot = pydot

ModuleNotFoundError: No module named 'pydot'

In [5]:
# Import the plotting function
from keras.utils import plot_model
import matplotlib.pyplot as plt


# Summarize the model
model.summary()

# Plot the model
plot_model(model, to_file='model.png')

# Display the image
data = plt.imread('model.png')
plt.imshow(data)
plt.show()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_3 (InputLayer)         (None, 1)                 0         
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 2         
Total params: 2
Trainable params: 2
Non-trainable params: 0
_________________________________________________________________


ImportError: Failed to import `pydot`. Please install `pydot`. For example with `pip install pydot`.

## Fit and evaluate a model

<img src="images/19.PNG" style="width:650px;height:300px;"> 

<img src="images/20.PNG" style="width:650px;height:300px;"> 

<img src="images/21.PNG" style="width:650px;height:300px;"> 

<img src="images/22.PNG" style="width:650px;height:300px;"> 

<img src="images/23.PNG" style="width:650px;height:300px;"> 

<img src="images/24.PNG" style="width:650px;height:300px;"> 

<img src="images/25.PNG" style="width:650px;height:300px;"> 

<img src="images/26.PNG" style="width:650px;height:300px;"> 

In [54]:
import pandas as pd
games_games_season = pd.read_csv('datasets/games_season.csv')
games_games_season.head()

season  team_1  team_2  home  score_diff  score_1  score_2  won
0    1985    3745    6664     0          17       81       64    1
1    1985     126    7493     1           7       77       70    1
2    1985     288    3593     1           7       63       56    1
3    1985    1846    9881     1          16       70       54    1
4    1985    2675   10298     1          12       86       74    1

In [46]:
games_tourney = pd.read_csv('datasets/games_tourney.csv')
games_tourney.head()

season  team_1  team_2  home  seed_diff  score_diff  score_1  score_2  won
0    1985     288      73     0         -3          -9       41       50    0
1    1985    5929      73     0          4           6       61       55    1
2    1985    9884      73     0          5          -4       59       63    0
3    1985      73     288     0          3           9       50       41    1
4    1985    3920     410     0          1          -9       54       63    0

## Build the model

In [47]:
from keras.models import Model
from keras.layers import Input, Dense
input_tensor = Input(shape=(1,))
output_tensor = Dense(1)(input_tensor)
model = Model(input_tensor, output_tensor)
model.compile(optimizer='adam', loss='mae')

## Fit the model

 - Fit the model to the tournament basketball data

 - Now that the model is compiled, you are ready to fit it to some data!

 - In this exercise, you'll use a dataset of scores from US College Basketball tournament games. Each row of the dataset has the team ids: team_1 and team_2, as integers. It also has the seed difference between the teams (seeds are assigned by the tournament committee and represent a ranking of how strong the teams are) and the score difference of the game (e.g. if team_1 wins by 5 points, the score difference is 5).

 - To fit the model, you provide a matrix of X variables (in this case one column: the seed difference) and a matrix of Y variables (in this case one column: the score difference).

 - The games_tourney DataFrame along with the compiled model object is available in your workspace.
 

## Instructions

   - Fit the model with seed_diff as the input variable and score_diff as the output variable.
   - Use 1 epoch, a batch size of 128, and a 10% validation split.


In [48]:
# Now fit the model
model.fit(games_tourney['seed_diff'], games_tourney['score_diff'],
          epochs=1,
          batch_size=128,
          validation_split=0.10,
          verbose=True)

Train on 3810 samples, validate on 424 samples
Epoch 1/1
3810/3810 [==============================] - 0s 57us/step - loss: 9.2000 - val_loss: 8.8008


## Evaluate the model on a test set

 - After fitting the model, you can evaluate it on new data. You will give the model a new X matrix (also called test data), allow it to make predictions, and then compare to the known y variable (also called target data).

 - In this case, you'll use data from the post-season tournament to evaluate your model. The tournament games happen after the regular season games you used to train our model, and are therefore a good evaluation of how well your model performs out-of-sample.

 - The games_tourney_test DataFrame along with the fitted model object is available in your workspace.
 
 
### Instructions
 - Assign the test data (seed_diff column) to X_test.
 - Assign the target data (score_diff column) to y_test.
 - Evaluate the model on X_test and y_test.


In [49]:
# Load the X variable from the test data
X_test = games_tourney['seed_diff']

# Load the y variable from the test data
y_test = games_tourney['score_diff']

# Evaluate the model on the test data
model.evaluate(X_test, y_test)

4234/4234 [==============================] - 0s 10us/step


9.156618082247425

## 2 Two Input Networks Using Categorical Embeddings, Shared Layers, and Merge Layers

   - In this chapter, you will build two-input networks that use categorical embeddings to represent high-cardinality data, shared layers to specify re-usable building blocks, and merge layers to join multiple inputs to a single output. By the end of this chapter, you will have the foundational building blocks for designing neural networks with complex data flows.

 - However, we have a much bigger dataset with over 300,000 regular seasons games 

## category embedding 

<img src="images/27.PNG" style="width:650px;height:300px;"> 




 - **Categorical Embedding** are an **advanced type of layer**, only available in Deep Learning Libraries.
 - They are extremely useful for **dealing with high cardinality categorical data**.
 - In this data set the team ID variable has **high cardinality**.
 - Embedding layers are also useful dealing with text data, such as **Word2Vec** models.
 
 
To Map the integer team ID's to a decimal rating, we use an embedding layer.

## To get started with category embeddings, you will need an input layer.
   - In this case, your input is a single number, ranging from 1 to 10,887, which represent each team unique id.
   - data set has about 400 unique schools, giving us close to 12,000 ID's.
   - **To Create a embedding layer, use the Embedding() function from the keras.** 
   

<img src="images/28.PNG" style="width:650px;height:300px;"> 




<img src="images/29.PNG" > 




## Embedding layers increase the dimensionality of your data.
 - the input CSV has two dimensions(row and columns), But embedding layers add a third dimension.
 - This third dimension can be useful when dealing with images and text.
 - So we use the **flatten layer to flatten the embedding from 3D to 2D**.
 - The flatten layer is also the output layer for the embedding process.
 - **Flatten layers are useful for dealing with time series data, Text data and images**.
 

<img src="images/30.PNG"> 




## Shared layers

## Define team lookup

 - **Shared layers allow a model to use the same weight matrix for multiple steps**. 
 - In this exercise, you will build a "team strength" layer that represents each team by a single number. 
 - You will use this number for both teams in the model. The model will learn a number for each team that works well both when the team is team_1 and when the team is team_2 in the input data.

The games_season DataFrame is available in your workspace.

## Instructions

   - Count the number of unique teams.
   - Create an embedding layer that maps each team ID to a single number representing that team's strength.
   - The output shape should be 1 dimension (as we want to represent the teams by a single number).
   - The input length should be 1 dimension (as each team is represented by exactly one id).


In [6]:
import pandas as pd
games_season = pd.read_csv('datasets/games_season.csv')
games_season.head()

season  team_1  team_2  home  score_diff  score_1  score_2  won
0    1985    3745    6664     0          17       81       64    1
1    1985     126    7493     1           7       77       70    1
2    1985     288    3593     1           7       63       56    1
3    1985    1846    9881     1          16       70       54    1
4    1985    2675   10298     1          12       86       74    1

In [7]:
# Imports
from keras.layers import Embedding
from numpy import unique

# Count the unique number of teams
n_teams = unique(games_season[['team_1','team_2']]).shape[0]
print(n_teams)
# Create an embedding layer
team_lookup = Embedding(input_dim=n_teams,
                        output_dim=1,
                        input_length=1,
                        name='Team-Strength')

10888


## The embedding layer is a lot like a dictionary, but your model learns the values for each key.

## Define team model

 - The team strength lookup has three components: an input, an embedding layer, and a flatten layer that creates the output.

 - If you wrap these three layers in a model with an input and output, **you can re-use that stack of three layers at multiple places**.

 - Note again that the weights for all three layers will be shared everywhere we use them.

## Instructions
 - Create a 1D input layer for the team ID (which will be an integer). Be sure to set the correct input shape!
 - Pass this input to the team strength lookup layer you created previously.
 - Flatten the output of the team strength lookup.
 - Create a model that uses the 1D input as input and flattened team strength as output.


In [52]:
# Imports
from keras.layers import Input, Embedding, Flatten
from keras.models import Model

# Create an input layer for the team ID
teamid_in = Input(shape=(1,))

# Lookup the input in the team strength embedding layer
strength_lookup = team_lookup(teamid_in)

# Flatten the output
strength_lookup_flat = Flatten()(strength_lookup)

# Combine the operations into a single, re-usable model
team_strength_model = Model(teamid_in, strength_lookup_flat, name='Team-Strength-Model')

## The model will be reusable, so you can use it in two places in your final model.

## 2.1 Shared layer
In this chapter, you will create a model with two inputs: one for each team in the basketball dataset
   - However, you want these two teams to each use the same embedding layer you defined in the previous lesson.
   - **Accomplishing this requires a shared layer**.
   - ** Shared layers are an advanced Deep learning concept, and are only possible with the keras functional API**.
   - they allow you to define an operation and then apply the exact same operation(with exact same weight) on differnt inputs.
   - In this model, we will share team rating for both inputs.
   - The learned rating will be same, whether it applies to team1 ot team2.
   - ** To create a shared layer, you must first create two inputs, each will be passed to the shared layers**.
   

<img src="images/31.PNG" > 




<img src="images/32.PNG" > 




<img src="images/33.PNG" > 




<img src="images/34.PNG" > 




## Defining two inputs
 - In this exercise, you will define two input layers for the two teams in your model. This allows you to specify later in the model how the data from each team will be used differently.
 
 
## Instructions
  - Create an input layer to use for team 1. Recall that our input dimension is 1.
  - Name the input "Team-1-In" so you can later distinguish it from team 2.
  - Create an input layer to use for team 2, named "Team-2-In".


In [53]:
# Load the input layer from keras.layers
from keras.layers import Input

# Input layer for team 1
team_in_1 = Input(shape=(1,), name='Team-1-In')

# Separate input layer for team 2
team_in_2 = Input(shape=(1,), name='Team-2-In')

## Lookup both inputs in the same model

 - Now that you have a team strength model and an input layer for each team, you can lookup the team inputs in the shared team strength model. The two inputs will share the same weights.

 - In this dataset, you have 10,888 unique teams. You want to learn a strength rating for each team, such that if any pair of teams plays each other, you can predict the score, even if those two teams have never played before. Furthermore, you want the strength rating to be the same, regardless of whether the team is the home team or the away team.

 - To achieve this, you use a shared layer, defined by the re-usable model (team_strength_model()) you built in exercise 3 and the two input layers (team_in_1 and team_in_2) from the previous exercise, all of which are available in your workspace.

In [54]:
# Lookup team 1 in the team strength model
team_1_strength = team_strength_model(team_in_1)

# Lookup team 2 in the team strength model
team_2_strength = team_strength_model(team_in_2)

## Merge Layers
Now that you've got multiple inputs and a shared layer, you need to combine your inputs into single layer that you can use to pred a single output.  

## This require a Merge layers

 - Merge layers allow you to define advanced, non-sequential network topologies.
 - There are many kinds of merge layers available in Keras Add, Subtract, Multiply, Concatenate.
 - Require them to be same shape.
 


<img src="images/35.PNG" > 




<img src="images/36.PNG" > 




<img src="images/37.PNG" > 




## Output layer using shared layer

 - Now that you've looked up how "strong" each team is, subtract the team strengths to determine which team is expected to win the game.

 - This is a bit like the seeds that the tournament committee uses, which are also a measure of team strength. But rather than using seed differences to predict score differences, you'll use the difference of your own team strength model to predict score differences.

 - The subtract layer will combine the weights from the two layers by subtracting them.

## Instructions
 - Import the Subtract layer from keras.layers.
 - Combine the two-team strength lookups you did earlier.


In [55]:
# Import the Subtract layer from keras
from keras.layers import Subtract

# Create a subtract layer using the inputs from the previous exercise
score_diff = Subtract()([team_1_strength, team_2_strength])

## Model using two inputs and one output

Now that you have your two inputs (team id 1 and team id 2) and output (score difference), you can wrap them up in a model so you can use it later for fitting to data and evaluating on new data.

## Instructions
   - Define a model with the two teams as inputs and use the score difference as the output.
   - Compile the model with the 'adam' optimizer and 'mean_absolute_error' loss.


In [56]:
# Imports
from keras.layers import Subtract
from keras.models import Model

# Subtraction layer from previous exercise
score_diff = Subtract()([team_1_strength, team_2_strength])

# Create the model
model = Model([team_in_1, team_in_2], score_diff)

# Compile the model
model.compile(optimizer='adam', loss='mean_absolute_error')

## Predict from your model
Keras models with multiple inputs work just like Keras models with single input.
 - They use the same fit, Evalaute, and predict methods.
 - The only differenrce is that all of those methods take lists of inputs rather then a single input.
 - 

<img src="images/38.PNG" > 




<img src="images/39.PNG" > 




<img src="images/40.PNG" > 

## Fit the model to the regular season training data


 - Now that you've defined a complete team strength model, you can fit it to the basketball data! Since your model has two inputs now, you need to pass the input data as a list.
 
 
## Instructions
 - Assign the 'team_1' and 'team_2' columns from games_season to input_1 and input_2, respectively.
 - Use 'score_diff' column from games_season as the target.
 - Fit the model using 1 epoch, a batch size of 2048, and a 10% validation split.


In [57]:
# Get the team_1 column from the regular season data
input_1 = games_season['team_1']

# Get the team_2 column from the regular season data
input_2 = games_season['team_2']

# Fit the model to input 1 and 2, using score diff as a target
model.fit([input_1, input_2],
          games_season['score_diff'],
          epochs=1,
          batch_size=2048,
          validation_split=0.10,
          verbose=True)

Train on 280960 samples, validate on 31218 samples
Epoch 1/1
280960/280960 [==============================] - 0s 2us/step - loss: 12.1201 - val_loss: 11.8385


## Evaluate the model on the tournament test data

 - The model you fit to the regular season data (model) in the previous exercise and the tournament dataset (games_tourney) are available in your workspace.

 - In this exercise, you will evaluate the model on this new dataset. This evaluation will tell you how well you can predict the tournament games, based on a model trained with the regular season data. This is interesting because many teams play each other in the tournament that did not play in the regular season, so this is a very good check that your model is not overfitting.

In [58]:
# Get team_1 from the tournament data
input_1 = games_tourney['team_1']

# Get team_2 from the tournament data
input_2 = games_tourney['team_2']

# Evaluate the model using these inputs
model.evaluate([input_1, input_2], games_tourney['score_diff'])

4234/4234 [==============================] - 0s 12us/step


11.680247329347923

## 3 Multiple Inputs: 3 Inputs (and Beyond!)

 - In this chapter, you will extend your 2-input model to 3 inputs, and learn how to use Keras' **summary and plot functions** to understand the parameters and topology of your neural networks. By the end of the chapter, you will understand how to extend a **2-input model to 3 inputs and beyond**.

## Three-input models
Once you have learned how to work with two input networks,it is trivial to extend that knowledge to 3 or more input network
   - Making a keras model with 3 input is almost exactly same as making a keras model with two inputs.
     

<img src="images/41.PNG" > 

<img src="images/42.PNG" > 

<img src="images/43.PNG" > 

<img src="images/44.PNG" > 

<img src="images/45.PNG" > 

## Make an input layer for home vs. away


 - Now you will make an improvement to the model you used in the previous chapter for regular season games. You know there is a well-documented home-team advantage in basketball, so you will add a new input to your model to capture this effect.

 - This model will have **three inputs: team_id_1, team_id_2, and home**. The team IDs will be integers that you look up in your team strength model from the previous chapter, and home will be a binary variable, 1 if team_1 is playing at home, 0 if they are not.

 - The team_strength_model you used in the previous chapter has been loaded into your workspace. After applying it to each input, use a Concatenate layer to join the two team strengths and with the home vs away variable, and pass the result to a Dense layer.
 

## Instructions

 - Create three inputs layers of shape 1, one each for team 1, team 2, and home vs away.
 - Lookup the team inputs in team_strength_model().
 - Concatenate the team strengths with the home input and pass to a Dense layer.

In [59]:
from keras.layers import concatenate
# Create an Input for each team
team_in_1 = Input(shape=(1,), name='Team-1-In')
team_in_2 = Input(shape=(1,), name='Team-2-In')

# Create an input for home vs away
home_in = Input(shape=(1,), name='Home-In')

# Lookup the team inputs in the team strength model
team_1_strength = team_strength_model(team_in_1)
team_2_strength = team_strength_model(team_in_2)

# Combine the team strengths with the home input using a Concatenate layer, then add a Dense layer
out = concatenate([team_1_strength, team_2_strength, home_in])
out = Dense(1)(out)

### Make a model and compile it
 - Now that you've input and output layers for the 3-input model, wrap them up in a Keras model class, and then compile the model, so you can fit it to data and use it to make predictions on new data.

## Instructions
 - Create a model using team_in_1, team_in_2, and home_in as inputs and out as the output.
 - Compile the model using the 'adam' optimizer and 'mean_absolute_error' as the loss function.

In [60]:
# Import the model class
from keras.models import Model

# Make a Model
model = Model([team_in_1, team_in_2, home_in], out)

# Compile the model
model.compile(optimizer='adam', loss='mean_absolute_error')

## Fit the model and evaluate
 - Now that you've defined a new model, fit it to the regular season basketball data.
 - Use the model you fit in the previous exercise (which was trained on the regular season data) and evaluate the model on data for tournament games (games_tourney).

## Instructions

 - Fit the model to the games_season dataset, using 'team_1', 'team_2' and 'home' columns as inputs, and the 'score_diff' column as the target.
 - Fit the model using 1 epoch, 10% validation split and a batch size of 2048.
 - Evaluate the model on games_tourney, using the same inputs and outputs.

In [61]:
# Fit the model to the games_season dataset
model.fit([games_season['team_1'], games_season['team_2'], games_season['home']],
          games_season['score_diff'],
          epochs=1,
          verbose=True,
          validation_split=.10,
          batch_size=2048)

# Evaluate the model on the games_tourney dataset
model.evaluate([games_tourney['team_1'], games_tourney['team_2'], games_tourney['home']],
               games_tourney['score_diff'])

Train on 280960 samples, validate on 31218 samples
Epoch 1/1
4234/4234 [==============================] - 0s 11us/step


11.677241001931012

## Summarizing and plotting models


The summary shows you all the layers in the model, as well as how many parameters each layer has.
 - Importantly, Keras models can have non-trainable parameters that are fixed and do not change, as well as trainable parameters, that are learned from the data when the model is fit.
 - Model with more trainable parameters are typicall more flexible.
 - This can also make them more prone to overfitting.
 - models with fewer trainable parameters are less flexible, but therefore less likely to overfit.
 - In this case the model has three inputs, since all three of them feed into one Dense layer, The model has four parameter(Input+ Bias).
 - 

<img src="images/46.PNG" > 

## Slightly more complicated model

<img src="images/47.PNG" > 

<img src="images/48.PNG" > 

## Model summaries

 - In this exercise, you will take a closer look at the summary of one of your **3-input models** available in your workspace as model. Note how many layers the model has, how many parameters it has, and how many of those parameters are trainable/non-trainable.


In [62]:
model.summary()

__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
Team-1-In (InputLayer)          (None, 1)            0                                            
__________________________________________________________________________________________________
Team-2-In (InputLayer)          (None, 1)            0                                            
__________________________________________________________________________________________________
Team-Strength-Model (Model)     (None, 1)            10888       Team-1-In[0][0]                  
                                                                 Team-2-In[0][0]                  
__________________________________________________________________________________________________
Home-In (InputLayer)            (None, 1)            0                                            
__________

## Plotting models

 - In addition to summarizing your model, you can also plot your model to get a more intuitive sense of it. Your model is available in the workspace.
 

In [63]:
# Imports
import matplotlib.pyplot as plt
from keras.utils import plot_model

# Plot the model
plot_model(model, to_file='model.png')

# Display the image
data = print(to_file)

plt.show()

ImportError: Failed to import `pydot`. Please install `pydot`. For example with `pip install pydot`.

## Stacking models

"Stacking" or using the predictions from one mnodel as an input to another model
 - **Model stacking is a very advanced data science concept, it is the most sophisticated way of combining models, and when doen right can yield some of the most accurate models in existance**

## datasets:
 - The collage basketball data from the regular season and the collage basketball from the post-season tournament.
 - Both the datasets contain the two teams playing, whether **team 1** is home or away and the score difference of the games.
 - the tournament dataset additionally contain the difference in seeds of the two teams playing.



## Stacking model
 - You start with the regular season dataset and fit a model to it.
 - You then predict on the tournament dataset, using this model.
 - this gives you predicted tournament outcomes, which you can now use to build model of the actual tournament outcomes.
 - The predictions from the regular season model capture the effect of team_1 and team_2, which means you now don't need to use those two variables in the tournament model, and can avoid the use of an embedding layer.
 
<img src="images/49.PNG" > 

**The only draw back of this approach is that all the inputs must be numeric**

<img src="images/50.PNG" > 

## Add the model predictions to the tournament data

 - In lesson 1 of this chapter, you used the regular season model to make predictions on the tournament dataset, and got pretty good results! Try to improve your predictions for the tournament by modeling it specifically.

 - You'll use the prediction from the regular season model as an input to the tournament model. This is a form of "model stacking."

 - To start, take the regular season model from the previous lesson, and predict on the tournament data. Add this prediction to the tournament data as a new column.
 
 
## Instructions

 - Use the model to predict on the games_tourney dataset. The model has three inputs: 'team_1', 'team_2', and 'home' columns.     - Assign the predictions to a new column, 'pred'.



In [64]:
# Predict
games_tourney['pred'] = model.predict([games_tourney['team_1'], 
                                             games_tourney['team_2'], 
                                             games_tourney['home']])

## Create an input layer with multiple columns

 - In this exercise, you will look at a different way to **create models with multiple inputs**. 
 - **This method only works for purely numeric data, but its a much simpler approach to making multi-variate neural networks**.
 - Now you have three numeric columns in the tournament dataset:**'seed_diff', 'home', and 'pred'**. 
 - In this exercise, you will create a neural network that uses a **single input layer to process all three of these numeric inputs**.
 - This model should have a **single output to predict the tournament game score difference**.
 

## Instructions

 - Create a single input layer with 3 columns.
 - Connect this input to a Dense layer with 1 unit.
 - Create a model with input_tensor as the input and output_tensor as the output.
 - Compile the model with 'adam' as the optimizer and 'mean_absolute_error' as the loss function.

In [65]:
# Create an input layer with 3 columns
input_tensor = Input((3,))

# Pass it to a Dense layer with 1 unit
output_tensor = Dense(1)(input_tensor)

# Create a model
model = Model(input_tensor, output_tensor)

# Compile the model
model.compile(optimizer='adam', loss='mean_absolute_error')

## Fit the model

 - Now that you've enriched the tournament dataset and built a model to make use of the new data, fit that model to the tournament data.

 - Note that this model has only one input layer that is capable of handling all 3 inputs, so it's inputs and outputs do not need to be a list.

 - Tournament games are split into a training set and a test set. **The tournament games before 2010 are in the training set**, and the **ones after 2010 are in the test set**.

## Instructions

- Fit the model to the games_tourney_train dataset using 1 epoch.
- The input columns are 'home', 'seed_diff', and 'pred'.
- The target column is 'score_diff'.

In [66]:
games_tourney.head()

season  team_1  team_2  home  seed_diff  score_diff  score_1  score_2  won  \
0    1985     288      73     0         -3          -9       41       50    0   
1    1985    5929      73     0          4           6       61       55    1   
2    1985    9884      73     0          5          -4       59       63    0   
3    1985      73     288     0          3           9       50       41    1   
4    1985    3920     410     0          1          -9       54       63    0   

       pred  
0  0.163612  
1  0.200548  
2  0.191497  
3  0.132147  
4  0.157973

In [79]:
games_tourney_train = games_tourney[games_tourney['season'] < 2010]
games_tourney_train.tail()  

        

season  team_1  team_2  home  seed_diff  score_diff  score_1  score_2  \
3163    2009    2902   10688     0          7          -2       59       61   
3164    2009   10810   10688     0          8          11       60       49   
3165    2009    7024   10810     0          3           5       60       55   
3166    2009    7078   10810     0         -9         -18       59       77   
3167    2009   10688   10810     0         -8         -11       49       60   

      won      pred  
3163    0  0.166632  
3164    1  0.203057  
3165    1  0.158334  
3166    0  0.111178  
3167    0  0.098952

In [83]:
games_tourney_test = games_tourney[games_tourney['season'] >= 2010]
games_tourney_test.head()

season  team_1  team_2  home  seed_diff  score_diff  score_1  score_2  \
3168    2010    2365     401     0         15          29       73       44   
3169    2010   10655     401     0          0         -17       44       61   
3170    2010    2365     647     0          2           7       78       71   
3171    2010    6757     647     0         -8          -8       68       76   
3172    2010    7616     647     0        -11          -9       59       68   

      won      pred  
3168    1  0.223730  
3169    0  0.098897  
3170    1  0.196799  
3171    0  0.161127  
3172    0  0.134733

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df, y, test_size=0.2)
print X_train.shape, y_train.shape
print X_test.shape, y_test.shape

In [80]:
# Fit the model
model.fit(games_tourney_train[['home', 'seed_diff','pred']],
          games_tourney_train['score_diff'],
          epochs=1,
          verbose=True)

Epoch 1/1
3168/3168 [==============================] - 0s 77us/step - loss: 14.2156


In [85]:
# Evaluate the model on the games_tourney_test dataset
model.evaluate(games_tourney_test[['home', 'seed_diff', 'pred']], 
               games_tourney_test['score_diff'])

1066/1066 [==============================] - 0s 82us/step


13.110603488184944

## 4 Multiple Outputs

In this chapter, you will build neural networks with multiple outputs, which can be used to solve regression problems with multiple targets. You will also build a model that solves a regression problem and a classification problem simultaneously.


 - we'll cover neural networks with 2 outputs make predictions for2 target at once.
 - You could use a single model to predict the scores of both teams in a basketball or .
 - Use a single model to predict both the score difference and the Win/Loss outcome of that Game.

<img src="images/51.PNG" > 

<img src="images/52.PNG" > 

<img src="images/53.PNG" > 

<img src="images/54.PNG" > 

<img src="images/55.PNG" > 

## Simple two-output model

- In this exercise, you will use the tournament data to build one model that makes two predictions: the scores of both teams in a given game. 
- Your **inputs will be the seed difference of the two teams**, as well as the predicted score difference from the model you built in chapter 3.

- The output from your model will be the predicted score for team 1 as well as team 2. This is called "**multiple target regression**": one model making more than one prediction.


## Instructions

 - Create a single input layer with 2 columns.
 - Connect this input to a Dense layer with 2 units.
 - Create a model with input_tensor as the input and output_tensor as the output.
 - Compile the model with 'adam' as the optimizer and 'mean_absolute_error' as the loss function.

In [89]:
# Define the input
input_tensor = Input(shape=(2,))

# Define the output
output_tensor = Dense(2)(input_tensor)

# Create a model
model = Model(input_tensor, output_tensor)

# Compile the model
model.compile(loss='mean_absolute_error', optimizer='adam')

## Fit a model with two outputs


Now that you've defined your 2-output model, fit it to the tournament data. I've split the data into games_tourney_train and games_tourney_test, so use the training set to fit for now.

This model will use the pre-tournament seeds, as well as your pre-tournament predictions from the regular season model you built previously in this course.

As a reminder, this model will predict the scores of both teams.


## Instructions

Fit the model to the games_tourney_train dataset using 100 epochs and a batch size of 16384.
The input columns are 'seed_diff', and 'pred'.
The target columns are 'score_1' and 'score_2'.



In [90]:
model.fit(games_tourney_train[['seed_diff', 'pred']],
  		  games_tourney_train[['score_1', 'score_2']],
  		  verbose=True,
          epochs=100,
  		  batch_size=16384)

Epoch 1/100
3168/3168 [==============================] - 0s 72us/step - loss: 71.8957
Epoch 2/100
3168/3168 [==============================] - 0s 1us/step - loss: 71.8946
Epoch 3/100
3168/3168 [==============================] - 0s 1us/step - loss: 71.8934
Epoch 4/100
3168/3168 [==============================] - 0s 1us/step - loss: 71.8922
Epoch 5/100
3168/3168 [==============================] - 0s 1us/step - loss: 71.8911
Epoch 6/100
3168/3168 [==============================] - 0s 1us/step - loss: 71.8899
Epoch 7/100
3168/3168 [==============================] - 0s 1us/step - loss: 71.8888
Epoch 8/100
3168/3168 [==============================] - 0s 1us/step - loss: 71.8876
Epoch 9/100
3168/3168 [==============================] - 0s 1us/step - loss: 71.8864
Epoch 10/100
3168/3168 [==============================] - 0s 2us/step - loss: 71.8853
Epoch 11/100
3168/3168 [==============================] - 0s 1us/step - loss: 71.8841
Epoch 12/100
3168/3168 [==============================] - 0s 1

Epoch 97/100
3168/3168 [==============================] - 0s 1us/step - loss: 71.7847
Epoch 98/100
3168/3168 [==============================] - 0s 1us/step - loss: 71.7836
Epoch 99/100
3168/3168 [==============================] - 0s 1us/step - loss: 71.7824
Epoch 100/100
3168/3168 [==============================] - 0s 1us/step - loss: 71.7813


## Inspect the model (I)

 - Now that you've fit your model, let's take a look at it. You can use the .get_weights() method to inspect your model's weights.

 - The input layer will have 4 weights: 2 for each input times 2 for each output.

 - The output layer will have 2 weights, one for each output.

## Instructions

 - Print the model's weights.
 - Print the column means of the training data (games_tourney_train).

In [92]:
# Print the model's weights
print(model.get_weights())

# Print the column means of the training data
print(games_tourney_train.mean())

[array([[ 0.33546776, -0.44695818],
       [ 1.2787606 , -0.8660291 ]], dtype=float32), array([0.10004997, 0.10004997], dtype=float32)]
season        1997.045455
team_1        5546.025568
team_2        5546.025568
home             0.000000
seed_diff        0.000000
score_diff       0.000000
score_1         71.912247
score_2         71.912247
won              0.500000
pred             0.155596
dtype: float64


## Did you notice that both output weights are about ~72? This is because, on average, a team will score about 72 points in the tournament.

## Evaluate the model

 - Now that you've fit your model and inspected it's weights to make sure it makes sense, evaluate it on the tournament test set to see how well it performs on new data.

## Instructions

- Evaluate the model on games_tourney_test.
- Use the same inputs and outputs as the training set.

In [93]:
model.evaluate(games_tourney_test[['seed_diff', 'pred']],
               games_tourney_test[['score_1', 'score_2']])

1066/1066 [==============================] - 0s 100us/step


68.67861587782068

## Single model for classification and regression

<img src="images/57.PNG" > 



<img src="images/58.PNG" > 

<img src="images/59.PNG" > 

<img src="images/60.PNG" > 

<img src="images/61.PNG" > 

## Classification and regression in one model

Now you will create a different kind of 2-output model. This time, you will predict the score difference, instead of both team's scores and then you will predict the probability that team 1 won the game. This is a pretty cool model: it is going to do both classification and regression!

In this model, turn off the bias, or intercept for each layer. Your inputs (seed difference and predicted score difference) have a mean of very close to zero, and your outputs both have means that are close to zero, so your model shouldn't need the bias term to fit the data well.

## Instructions

Create a single input layer with 2 columns.

The first output layer should have 1 unit with 'linear' activation and no bias term.

The second output layer should have 1 unit with 'sigmoid' activation and no bias term. Also, use the first output layer as an input to this layer.

Create a model with these input and outputs.

In [94]:
# Create an input layer with 2 columns
input_tensor = Input(shape=(2,))

# Create the first output
output_tensor_1 = Dense(1, activation='linear', use_bias=False)(input_tensor)

# Create the second output (use the first output as input here)
output_tensor_2 = Dense(1, activation='sigmoid', use_bias=False)(output_tensor_1)

# Create a model with 2 outputs
model = Model(input_tensor, [output_tensor_1, output_tensor_2])

## Compile and fit the model

 - Now that you have a model with 2 outputs, compile it with 2 loss functions: mean absolute error (MAE) for 'score_diff' and binary cross-entropy (also known as logloss) for 'won'. Then fit the model with 'seed_diff' and 'pred' as inputs. For outputs, predict 'score_diff' and 'won'.

 - This model can use the scores of the games to make sure that close games (small score diff) have lower win probabilities than blowouts (large score diff).

 - The regression problem is easier than the classification problem because MAE punishes the model less for a loss due to random chance. For example, if score_diff is -1 and won is 0, that means team_1 had some bad luck and lost by a single free throw. The data for the easy problem helps the model find a solution to the hard problem.
 
 
## Instructions

- Import Adam from keras.optimizers.
- Compile the model with 2 losses: 'mean_absolute_error' and 'binary_crossentropy', and use the Adam optimizer with a learning rate of 0.01.
- Fit the model with 'seed_diff' and 'pred' columns as the inputs and 'score_diff' and 'won' columns as the targets.
- Use 10 epochs and a batch size of 16384.


In [95]:
# Import the Adam optimizer
from keras.optimizers import Adam

# Compile the model with 2 losses and the Adam optimzer with a higher learning rate
model.compile(loss=['mean_absolute_error', 'binary_crossentropy'], optimizer=Adam(.01))

# Fit the model to the tournament training data, with 2 inputs and 2 outputs
model.fit(games_tourney_train[['seed_diff', 'pred']],
          [games_tourney_train[['score_diff']], games_tourney_train[['won']]],
          epochs=10,
          verbose=True,
          batch_size=16384)

Epoch 1/10
3168/3168 [==============================] - 0s 115us/step - loss: 16.5395 - dense_14_loss: 9.1985 - dense_15_loss: 7.3410
Epoch 2/10
3168/3168 [==============================] - 0s 0us/step - loss: 16.4776 - dense_14_loss: 9.2046 - dense_15_loss: 7.2730
Epoch 3/10
3168/3168 [==============================] - 0s 0us/step - loss: 16.4183 - dense_14_loss: 9.2111 - dense_15_loss: 7.2072
Epoch 4/10
3168/3168 [==============================] - 0s 2us/step - loss: 16.3596 - dense_14_loss: 9.2178 - dense_15_loss: 7.1418
Epoch 5/10
3168/3168 [==============================] - 0s 1us/step - loss: 16.3011 - dense_14_loss: 9.2246 - dense_15_loss: 7.0765
Epoch 6/10
3168/3168 [==============================] - 0s 2us/step - loss: 16.2432 - dense_14_loss: 9.2317 - dense_15_loss: 7.0114
Epoch 7/10
3168/3168 [==============================] - 0s 1us/step - loss: 16.1787 - dense_14_loss: 9.2394 - dense_15_loss: 6.9393
Epoch 8/10
3168/3168 [==============================] - 0s 1us/step - loss

## Inspect the model (II)

Now you should take a look at the weights for this model. In particular, note the last weight of the model. This weight converts the predicted score difference to a predicted win probability. If you multiply the predicted score difference by the last weight of the model and then apply the sigmoid function, you get the win probability of the game.


In [96]:
# Print the model weights
print(model.get_weights())

# Print the training data means
print(games_tourney_train.mean())

[array([[ 0.9774287],
       [-1.1267143]], dtype=float32), array([[-1.4007249]], dtype=float32)]
season        1997.045455
team_1        5546.025568
team_2        5546.025568
home             0.000000
seed_diff        0.000000
score_diff       0.000000
score_1         71.912247
score_2         71.912247
won              0.500000
pred             0.155596
dtype: float64


## Instructions 2/2

- Print the approximate win probability predicted for a close game (1 point difference).
- Print the approximate win probability predicted blowout game (10 point difference).

In [97]:
# Import the sigmoid function from scipy
from scipy.special import expit as sigmoid

# Weight from the model
weight = 0.14

# Print the approximate win probability of a predicted close game
print(sigmoid(1 * weight))

# Print the approximate win probability of a predicted blowout game
print(sigmoid(10 * weight))

0.5349429451582145
0.8021838885585818


- So sigmoid(1 * 0.14) is 0.53, which represents a pretty close game and sigmoid(10 * 0.14) is 0.80, which represents a pretty likely win. In other words, if the model predicts a win of 1 point, it is less sure of the win than if it predicts 10 points. Who says neural networks are black boxes?

## Evaluate on new data with two metrics


Now that you've fit your model and inspected its weights to make sure they make sense, evaluate your model on the tournament test set to see how well it does on new data.

Note that in this case, Keras will return 3 numbers: the first number will be the sum of both the loss functions, and then the next 2 numbers will be the loss functions you used when defining the model.

## Instructions

Evaluate the model on games_tourney_test.
Use the same inputs and outputs as the training set.

In [98]:
# Evaluate the model on new data
model.evaluate(games_tourney_test[['seed_diff', 'pred']],
               [games_tourney_test[['score_diff']], games_tourney_test[['won']]])

1066/1066 [==============================] - 0s 145us/step


[15.263273561202116, 9.091030417866376, 6.1722432533750835]

<img src="images/62.PNG" > 

<img src="images/63.PNG" > 